### PR TITLE
Add viz_expansions parameter for debug

### PR DESCRIPTION
**Pull request by doisyg**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Dexory sim |

---

## Description of contribution in a few bullet points

This is a draft adding a `viz_expansions` parameter which when set to true publishes on the topic `/expansions` the path search for the SmacHybrid planner.
This is what I use for debugging planner issues, similarly (and up to date) to what is described in **https://github.com/ros-planning/navigation2/tree/main/nav2_smac_planner** readme.
If there is an interest, I can extend to other Smac planners and make it a proper PR.
WDYT ?

## Description of documentation updates required from your changes

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_smac_planner/include/nav2_smac_planner/a_star.hpp [hunk 1]
```diff
@@ -22,6 +22,7 @@
 #include <memory>
 #include <queue>
 #include <utility>
+#include <tuple>
```
Comment (SteveMacenski) [inline nav2_smac_planner/include/nav2_smac_planner/a_star.hpp:25]: You can remove the last paragraph and the code snippet from the readme if this is added.

Usual stuff on docs updates + the smac planner readme has a param guide inline, so updating that too. 

I imagine this makes things much slower, it would be good to have that boldly spelled out that this should never be enabled outside of debugging

File: nav2_smac_planner/src/smac_planner_hybrid.cpp [hunk 2]
```diff
@@ -337,6 +351,10 @@ nav_msgs::msg::Path SmacPlannerHybrid::createPlan(
   NodeHybrid::CoordinateVector path;
   int num_iterations = 0;
 
+  std::shared_ptr<std::vector<std::tuple<float, float>>> expansions = nullptr;
+  if (_viz_expansions) {
+    expansions = std::make_shared<std::vector<std::tuple<float, float>>>();
```
[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_smac_planner/src/smac_planner_hybrid.cpp:356]: Also, a unique pointer that you `.get()` on will reduce some compute time on reference counter semantics and pointer overhead. `.get()` will still be nullptr if its not allocated I believe

**Comment by SteveMacenski** [comment]
It could be added to others, but it would be of diminishing value. The 2D planner is very straightforward and stupid, so I wouldn't think there would be anything interesting to see there other than a sea of red arrows (though we could definitely do it).

The Lattice planner has non-trivially long primitives that this would only show the final end-point of them, so I think it would be rather hard to follow to be meaningful for debugging.

File: nav2_smac_planner/src/smac_planner_hybrid.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (doisyg) [reply nav2_smac_planner/src/smac_planner_hybrid.cpp]: > You don't seem to pass this to the function?

My bad, bad merge when switching from Humble to main.
Side note, I noticed tolerance is now set to 0 in: 
https://github.com/ros-planning/navigation2/blob/85fd06f20413e348059205a84ecb50765fc8f0c7/nav2_smac_planner/src/smac_planner_hybrid.cpp#L341
Is this voluntary and handled elsewhere ? Or does it means that we don't support planning with tolerance in the SmacHybrid planner ?




> Also, a unique pointer that you `.get()` on will reduce some compute time on reference counter semantics and pointer overhead. `.get()` will still be nullptr if its not allocated I believe

Done

[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_smac_planner/src/smac_planner_hybrid.cpp]: The feasible planners don't have margin for error since "close enough" can't know if its kinematically achievable. anyway, the analytic expansions are what's getting the precise path to goal once within the general window of the goal so it should be exact if its possible at all.

File: nav2_smac_planner/include/nav2_smac_planner/a_star.hpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_smac_planner/include/nav2_smac_planner/a_star.hpp:25]: Still this stuff

**Comment by doisyg** [comment]
And docs: https://github.com/ros-planning/navigation.ros.org/pull/424

**Review by SteveMacenski — APPROVED** [review]
